### DISCUSSION TITLE
What about function name collisions?

### DISCUSSION THREAD

**Discussion by scottpersinger** [modelcontextprotocol/modelcontextprotocol#128] [General]
### Pre-submission Checklist

- [X] I have verified that this discussion would not be more appropriate as an issue in a specific repository
- [X] I have searched existing discussions to avoid duplicates

### Discussion Topic

Is there any guidance for how to avoid tool function name collisions? I note that the current GDrive tool exports a function called "search":

https://github.com/modelcontextprotocol/servers/blob/0e7d79c580da693091e7e9711a69132e85eb47c1/src/gdrive/index.ts#L133

which is HIGHLY likely to cause a collision down the road, haha. A related issue is that generic function names are likely to confuse your agent when they are mixed together. All the "search_file", "get_data", etc... functions are likely to only work reliably if your agent has that single tool attached.

My understanding is that function calling protocol at the moment doesn't support namespaces, so in the interim perhaps just providing guidance like: 

> Function names are recommended to include a prefix, like, "gdrive_search", rather than using more generic names.

would be a way to help ensure that tools can be used together by agents.

**Comment by jspahrsummers** [comment]
Yep, completely agreed. There's a somewhat related discussion here: https://github.com/modelcontextprotocol/specification/discussions/94

We may want to consider giving guidance to _clients_ that they automatically use the server name for namespacing on their end.

**Comment by slimslenderslacks** [comment]
@jspahrsummers when building an agent loop pre mcp, it was sometimes useful to do tool _selection_ before sampling.  I guess you could kind of think of it as un-discovering the tools that were not relevant in the current thread. Do you anticipate an analogous requirement that impacts the mcp spec? 
As an aside, it seems like 3.5 Sonnet is *really* quite good at dealing with pretty large numbers of available tools!

**Reply by jspahrsummers** [reply]
I don't know if it needs to be spec-ified, per se. I think we can give clients a lot of latitude in how they filter the tool list, like how you've done it.

**Comment by Mehdi-Bl** [comment]
Yes had this issue and ended up prefixing each MCP tool set of functions. Normalizing is important like you do in your code. But feel a lot building MCP servers right now, have not big exprience or care about code quality. IT JUST WORKS.
I see the current examples are not enforcing those good practices. There is the risk of conflict.
Also I noticed, the hints are not so easy to point the model. The worst to implement was diff edit with Claude as there is more than a way to do it. Also Claude is bad, as assume a lot he knows the exact file content while it may have changed.

**Comment by olaservo** [comment]
Hi, I added some meeting notes to this related discussion [here](https://github.com/modelcontextprotocol/modelcontextprotocol/discussions/94#discussioncomment-13516174) on other ideas for namespacing/avoiding collisions.

**Comment by localden** [comment]
We've limited Discussions in this repository to meeting notes and are closing this thread. To continue the conversation, pick a channel from [Contributor Communication](https://modelcontextprotocol.io/community/communication). Thanks for posting.
